azext_iot/tests/iothub/__init__.py

```diff
@@ -186,6 +182,92 @@ def generate_job_names(self, count=1):
             self.create_random_name(prefix=PREFIX_JOB, length=32) for i in range(count)
         ]
 
+    def _create_storage_account(self):
+        """
+        Create a storage account and container if a storage account was not created yet.
+        Populate the following variables if needed:
+          - storage_account_name
+          - storage_container
+          - storage_cstring
+        """
+        self.entity_rg = ENTITY_RG
+        self.storage_account_name = STORAGE_ACCOUNT_NAME
+        self.storage_container = STORAGE_CONTAINER
+        self.storage_cstring = None
+        if not settings.env.azext_iot_teststorageaccount:
+            storage_list = self.cmd(
+                'storage account list -g "{}"'.format(self.entity_rg)
+            ).get_output_in_json()
+
+            target_storage = None
+            for storage in storage_list:
+                if storage["name"] == self.storage_account_name:
+                    target_storage = storage
+                    break
+
+            if not target_storage:
+                self.cmd(
+                    "storage account create -n {} -g {} --tags iothub={}".format(
+                        self.storage_account_name, self.entity_rg, self.entity_name
+                    )
+                )
+            self._populate_storage_cstring()
+
+        # This won't overwrite or error if container is already created.
+        self.cmd(
+            "storage container create -n {} --connection-string '{}'".format(
+                self.storage_container, self.storage_cstring
+            ),
+        )
+
+    def _populate_storage_cstring(self):
+        """
+        Method to populate storage_cstring
+        """
+        if not self.storage_cstring:
+            self.storage_cstring = self.cmd(
+                "storage account show-connection-string -n {} -g {}".format(
+                    self.storage_account_name, self.entity_rg
+                )
+            ).get_output_in_json()["connectionString"]
+
+    def _delete_storage_account(self):
+        """
+        Delete the storage account if it was created.
+        """
+        if not settings.env.azext_iot_teststorageaccount:
+            self.cmd(
+                "storage account delete -n {} -g {} -y".format(
+                    self.storage_account_name, self.entity_rg
+                ),
+            )
+
+        elif not settings.env.azext_iot_teststorageaccount:
+            self.cmd(
+                "storage container delete -n {} --connection-string '{}'".format(
+                    self.storage_account_name, self.storage_cstring
+                ),
+            )
+
+    def _add_test_tag(self, test_tag):
```

Comment:
I'm seeing multiple similar implementations for this, can we refactor for more re-use? `az resource tag` can be used to apply tags generically to ARM resources.

Reply:
done